tests/load/transformations/test_incremental_transforms.py

```diff
@@ -0,0 +1,237 @@
+import pytest
+from typing import Any, Iterator
+import dlt
+from dlt.extract.hints import TDataItem
+from dlt.common.destination.dataset import SupportsReadableDataset
+
+from tests.load.utils import DestinationTestConfiguration
+from tests.load.transformations.utils import transformation_configs, setup_transformation_pipelines
+from dlt.pipeline.exceptions import PipelineNeverRan
+
+
+@pytest.fixture
+def inc_p() -> dlt.Pipeline:
+    pipeline = dlt.pipeline(
+        pipeline_name="incremental_pipeline",
+        destination="duckdb",
+        dataset_name="incremental_dataset",
+        dev_mode=True,
+    )
+    return pipeline
+
+
+@dlt.resource(table_name="items")
+def first_load() -> Iterator[TDataItem]:
+    yield from [
+        {"id": 1},
+        {"id": 2},
+        {"id": 3},
+    ]
+
+
+@dlt.resource(table_name="items")
+def inc_load() -> Iterator[TDataItem]:
+    yield from [
+        {"id": 4},
+        {"id": 5},
+        {"id": 6},
+    ]
+
+
+EXPECTED_TRANSFORMED_DATA_FIRST_LOAD = [
+    {"id": 1, "double_items": 2},
+    {"id": 2, "double_items": 4},
+    {"id": 3, "double_items": 6},
+]
+
+EXPECTED_TRANSFORMED_DATA_SECOND_LOAD = [
+    {"id": 1, "double_items": 2},
+    {"id": 2, "double_items": 4},
+    {"id": 3, "double_items": 6},
+    {"id": 4, "double_items": 8},
+    {"id": 5, "double_items": 10},
+    {"id": 6, "double_items": 12},
+]
+
+
+def _assert_transformed_data(inc_p: dlt.Pipeline, expected_data: list[Any]) -> None:
+    found_data = (
+        inc_p.dataset()
+        .transformed_items.select("id", "double_items")
+        .df()
+        .to_dict(orient="records")
+    )
+
+    # make all dict keys lowercase for our favorite snowflake
+    found_data = [{k.lower(): v for k, v in d.items()} for d in found_data]
+    sorted_found_data = sorted(found_data, key=lambda x: x["id"])
+
+    assert sorted_found_data == expected_data
+
+
+@pytest.mark.essential
+@pytest.mark.parametrize(
+    "destination_config",
+    transformation_configs(),
+    ids=lambda x: x.name,
+)
+@pytest.mark.parametrize("transformation_type", ["python", "sql"])
+def test_state_based_incremental_transform(
+    destination_config: DestinationTestConfiguration, transformation_type: Any
+) -> None:
+    """
+    Here we demonstrate how to use the resource state for transform incrementals
+    """
+    # get pipelines and populate fruit pipeline
+    inc_p, dest_p = setup_transformation_pipelines(destination_config, transformation_type)
+
+    @dlt.transformation(
+        transformation_type=transformation_type,
+        write_disposition="append",
+    )
+    def transformed_items(dataset: SupportsReadableDataset[Any], last_loaded_load_id: str) -> Any:
```

Comment:
1. would be good to have a user friendly type like `dlt.dataset`
2. `SupportsReadableDataset` hides many nice things people may use like destination_client and sql_client

Reply:
can you write a test where `last_loaded_load_id` is an optional argument that is injected from configuration?

Reply:
can we have a transfromation that uses a model and has merge write disposition with a primary key?

Reply:
I have added "dlt.dataset" as factory method and "dlt.Dataset" as an alias of the ReadableDBAPIDataset now, so it is like the Pipeline class and factory. Let me know if you think this works.

Reply:
I also modified one of the examples to take the load_id from config and added a fake incremental that works with merging.